aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -451,6 +451,29 @@ Tensor new_full(
   return at::full(size, fill_value, self.options().merge_in(options));
 }
 
+namespace {
+TensorOptions linspace_logspace_infer_options(
+    Scalar start,
+    Scalar end,
+    const TensorOptions& options) {
+  auto result_options = options;
+  if (start.isComplex() || end.isComplex()) {
+    // Since result_options.has_dtype() returns true (dtype is default type),
+    // even if the user hasn't specified the dtype.
+    // We just check to see if either `start` or `end` is complex,
+    // and if the `result_dtype` is not complex (be it default float type or
+    // user provided), we cast it to default complex dtype with a Warning!.
+    auto result_dtype = c10::typeMetaToScalarType(options.dtype());
+    if (!at::isComplexType(result_dtype)) {
+      TORCH_WARN(
+          "As either `start` or `stop` is complex, return type will be the default complex type.");
```

Comment:
Maybe we should add a cautionary note here "In future, we may throw an error when a non-complex dtype arg is passed as input along with complex valued start or end value."

Because, ideally we'd wanna throw an error for this scenario: `torch.linspace(3, 10j, steps=5, dtype=torch.float)`